network/network_windows.go

```diff
@@ -112,15 +115,169 @@ func (nm *networkManager) newNetworkImpl(nwInfo *NetworkInfo, extIf *externalInt
 	return nw, nil
 }
 
+// newNetworkImplHnsV2 creates a new container network for HNSv2.
+func (nm *networkManager) newNetworkImplHnsV2(nwInfo *NetworkInfo, extIf *externalInterface) (*network, error) {
+	// Initialize HNS network.
+	hnsNetwork := &hcn.HostComputeNetwork{
+		Name: nwInfo.Id,
+		Dns: hcn.Dns{
+			Domain:     nwInfo.DNS.Suffix,
+			ServerList: nwInfo.DNS.Servers,
+		},
+		Ipams: []hcn.Ipam{
+			hcn.Ipam{
+				Type: "Static",
+			},
+		},
+		SchemaVersion: hcn.SchemaVersion{
+			Major: 2,
+			Minor: 0,
+		},
+	}
+
+	// Set hcn network adaptor name policy
+	// FixMe: Find a better way to check if a nic that is selected is not part of a vSwitch
+	if !strings.HasPrefix(extIf.Name, "vEthernet") {
+		netAdapterNamePolicy, err := policy.GetHcnNetAdapterPolicy(extIf.Name)
+		if err != nil {
+			log.Printf("[net] Failed to serialize network adapter policy due to error: %v", err)
+			return nil, err
+		}
+
+		hnsNetwork.Policies = append(hnsNetwork.Policies, netAdapterNamePolicy)
+	}
+
+	// Set hcn subnet policy
+	var vlanid int
+	var subnetPolicy []byte
+	opt, _ := nwInfo.Options[genericData].(map[string]interface{})
+	if opt != nil && opt[VlanIDKey] != nil {
+		var err error
+		vlanID, _ := strconv.ParseUint(opt[VlanIDKey].(string), 10, 32)
+		subnetPolicy, err = policy.SerializeHcnSubnetVlanPolicy((uint32)(vlanID))
+		if err != nil {
+			log.Printf("[net] Failed to serialize subnet vlan policy due to error: %v", err)
+			return nil, err
+		}
+
+		vlanid = (int)(vlanID)
+	}
+
+	// Set network mode.
+	switch nwInfo.Mode {
+	case opModeBridge:
+		hnsNetwork.Type = hcn.L2Bridge
+	case opModeTunnel:
+		hnsNetwork.Type = hcn.L2Tunnel
+	default:
+		return nil, errNetworkModeInvalid
+	}
+
+	// Populate subnets.
+	for _, subnet := range nwInfo.Subnets {
+		hnsSubnet := hcn.Subnet{
+			IpAddressPrefix: subnet.Prefix.String(),
+			Routes: []hcn.Route{
+				hcn.Route{
+					NextHop: subnet.Gateway.String(),
+					DestinationPrefix: "0.0.0.0/0",
+				},
+			},
+		}
+
+		// Set the subnet policy
+		if vlanid > 0 {
+			hnsSubnet.Policies = append(hnsSubnet.Policies, subnetPolicy)
+		}
+
+		hnsNetwork.Ipams[0].Subnets = append(hnsNetwork.Ipams[0].Subnets, hnsSubnet)
+	}
+
+	// Create the HNS network.
+	log.Printf("[net] HostComputeNetwork Create: %+v", hnsNetwork)
+	hnsResponse, err := hnsNetwork.Create()
+	log.Printf("[net] HostComputeNetwork Create response: %+v err: %v.", hnsResponse, err)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the network object.
+	nw := &network{
+		Id:               nwInfo.Id,
+		HnsId:            hnsResponse.Id,
+		Mode:             nwInfo.Mode,
+		Endpoints:        make(map[string]*endpoint),
+		extIf:            extIf,
+		VlanId:           vlanid,
+		EnableSnatOnHost: nwInfo.EnableSnatOnHost,
+		NetNs:            nwInfo.NetNs,
+	}
+
+	globals, err := hcn.GetGlobals()
+	if err != nil || globals.Version.Major <= hcn.HNSVersion1803.Major {
+		// err would be not nil for windows 1709 & below
```

Comment:
10 seconds? do we want to really sleep for 10 seconds?

Reply:
This code was added by HNS team in our repo. We won't need this for hnsv2 path. checking with them.